### PR TITLE
Issue 4426 add sorting to mdm queries

### PR DESCRIPTION
**Pull request by jmarchionatto**

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/mdm/MdmLinkDaoJpaImpl.java [hunk 1]
```diff
@@ -193,48 +214,79 @@ public Page<MdmLink> search(MdmQuerySearchParameters theParams) {
...
-		return new PageImpl<>(typedQuery.setFirstResult(pageRequest.getOffset()).setMaxResults(pageRequest.getCount()).getResultList(),
-			PageRequest.of(pageRequest.getPage(), pageRequest.getCount()),
-			totalResults);
+	private List<Order> getOrderList(MdmQuerySearchParameters theParams, CriteriaBuilder criteriaBuilder, Root<MdmLink> from) {
+		if (CollectionUtils.isEmpty(theParams.getSort())) {
+			return Collections.emptyList();
+		}
+
+		List<Order> orderList = new ArrayList<>();
+		theParams.getSort().forEach(sortSpec -> {
```
Comment (lukedegruchy) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/mdm/MdmLinkDaoJpaImpl.java:284]: Nitpick:  why not: getSort()
.stream()
.map(sortSpec -> ....) 
.collect(Collectors.toUnmodifiedList())

File: hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/provider/MdmProviderQueryLinkR4Test.java [hunk 2]
```diff
@@ -72,26 +78,126 @@ public void before() throws Exception {
...
+		@Test
+		public void byCreatedDescending() {
+			Parameters result = (Parameters) myMdmProvider.queryLinks(null, null, null, null,
+				new UnsignedIntType(0), new UnsignedIntType(10), new StringType("-myCreated"),
+				myRequestDetails, new StringType("Patient"));
+
+			List<Parameters.ParametersParameterComponent> linkList = getParametersByName(result, "link");
+			assertThat(linkList, hasSize(3));
+
+			List<Long> createdDates = linkList.stream().map(this::extractCreated).collect(Collectors.toList());
```
Comment (lukedegruchy) [inline hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/provider/MdmProviderQueryLinkR4Test.java:110]: Nitpick:  Stream operations are easier to read if they're one per line ex:

linkList.stream()
.map(....)
.collect(.....)

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/mdm/MdmLinkDaoJpaImpl.java (hunk 1 above)
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/mdm/MdmLinkDaoJpaImpl.java:284]: Sure. Good idea

File: hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/provider/MdmProviderQueryLinkR4Test.java (hunk 2 above)
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/provider/MdmProviderQueryLinkR4Test.java:110]: I think it is a matter of use. I found code more readable using just a line for a simple method reference mapping

File: hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/MdmControllerSvcImpl.java [hunk 3]
```diff
@@ -214,7 +232,7 @@ public IBaseParameters submitMdmSubmitJob(List<String> theUrls,  IPrimitiveType<
 		if (theBatchSize != null && theBatchSize.getValue() !=null && theBatchSize.getValue().longValue() > 0) {
 			params.setBatchSize(theBatchSize.getValue().intValue());
 		}
-		ReadPartitionIdRequestDetails details= new ReadPartitionIdRequestDetails(null, RestOperationTypeEnum.EXTENDED_OPERATION_SERVER, null, null, null);
+		new ReadPartitionIdRequestDetails(null, RestOperationTypeEnum.EXTENDED_OPERATION_SERVER, null, null, null);
```
Comment (tadgh) [inline hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/MdmControllerSvcImpl.java:235]: question: Why is this here? It is not assigned to a variable, is it ever used?

**Review by tadgh — CHANGES_REQUESTED** [review]
Overall looks good minus one comment. I do have a request though: Can you add a test that exercises pagination in conjunction with sort? I don't see one here, and there is new paging code here that previously didn't exist.

File: hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/MdmControllerSvcImpl.java (hunk 3 above)
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/MdmControllerSvcImpl.java:235]: Don't know. Looks like an old leftover. Removing it. Nice catch!

**Review by tadgh — APPROVED** [review]

**Pull request merged**